### PR TITLE
Leap/dashboard callout contact

### PR DESCRIPTION
**Pull request by simleap**
- callout locations(province, district, commune, village)
- callout voice upload
- contact locations(province, district, commune, village)

screenshot: https://goo.gl/oBBfgm

### CODE REVIEW

**Comment by samnang** [comment]
```
Bundler could not find compatible versions for gem "aws-sdk-core":
  In snapshot (Gemfile.lock):
    aws-sdk-core (= 2.10.81)

  In Gemfile:
    active_elastic_job was resolved to 2.0.0, which depends on
      aws-sdk (~> 2) was resolved to 2.10.81, which depends on
        aws-sdk-resources (= 2.10.81) was resolved to 2.10.81, which depends on
          aws-sdk-core (= 2.10.81)

    aws-sdk-s3 was resolved to 1.10.0, which depends on
      aws-sdk-core (~> 3)

Running `bundle update` will rebuild your snapshot from scratch, using only
the gems in your Gemfile, which may resolve the conflict.

Bundler could not find compatible versions for gem "aws-sdk-s3":
  In Gemfile:
    aws-sdk-s3

Could not find gem 'aws-sdk-s3' in any of the sources.
```

It seems S3 is not working because of incompatible version with active_elastic_job, we need to patch this gem to support aws-sdk version 3

**Comment by samnang** [comment]
NOTE: We need to add this to the bucket to support directly uploads:

```xml
<?xml version="1.0" encoding="UTF-8"?>
<CORSConfiguration xmlns="http://s3.amazonaws.com/doc/2006-03-01/">
<CORSRule>
    <AllowedOrigin>https://*.mysite.com</AllowedOrigin>
    <AllowedMethod>GET</AllowedMethod>
    <MaxAgeSeconds>3000</MaxAgeSeconds>
    <AllowedHeader>Authorization</AllowedHeader>
</CORSRule>
<CORSRule>
    <AllowedOrigin>https://*.ews1294.info</AllowedOrigin>
    <AllowedMethod>PUT</AllowedMethod>
    <AllowedMethod>POST</AllowedMethod>
    <MaxAgeSeconds>3000</MaxAgeSeconds>
    <AllowedHeader>*</AllowedHeader>
</CORSRule>
</CORSConfiguration>
```

**Comment by simleap** [comment]
screenshots for callout voice validations: https://goo.gl/P6JA5R

**Comment by samnang** [comment]
@simleap anything left?

**Comment by simleap** [comment]
@samnang, it’s ready.

File: app/models/callout.rb [hunk 1]
```diff
@@ -63,4 +72,28 @@ class Callout < ApplicationRecord
...
   end
+
+  private
+
+  def validate_commune_ids
+    if commune_ids.blank? || commune_ids.reject!(&:empty?).blank?
+      errors.add(:commune_ids, :blank)
+    else
+      commune_ids.each do |commune_id|
+        next if commune_id =~ /^#{province_id}/
```
Comment (samnang) [inline app/models/callout.rb]: Shouldn't client side already handle this?

File: app/models/callout.rb [hunk 2]
```diff
@@ -63,4 +72,28 @@ class Callout < ApplicationRecord
   def title
     metadata["title"] || "Calout #{id}"
   end
+
+  private
+
+  def validate_commune_ids
+    if commune_ids.blank? || commune_ids.reject!(&:empty?).blank?
```
Comment (samnang) [inline app/models/callout.rb]: We should have `before_validation` to clean up empty strings, then the validation will be simpler.

```rb
before_validation :clean_up_blank_commune_ids

def clean_up_empty_commune_ids
  self.commune_ids = Array(commune_ids).reject(&:blank?)
end
```

File: app/models/callout.rb [hunk 3]
```diff
@@ -63,4 +72,28 @@ class Callout < ApplicationRecord
...
+    end
+  end
+
+  # https://github.com/rails/rails/issues/31656
+  def validate_voice
+    if voice.attached?
+      errors.add(:voice, :audio_type) unless voice.blob.audio?
+      errors.add(:voice, :audio_size) if voice.blob.byte_size.bytes > 5.megabytes
+    else
+      errors.add(:voice, :blank)
```
Comment (samnang) [inline app/models/callout.rb]: Can we just use `validate :voice, presence: true`?

File: config/locales/en.yml [hunk 4]
```diff
@@ -10,3 +10,6 @@ en:
           attributes:
             base:
               restrict_destroy_status: "Cannot delete record because the destroyer is not the creator"
+      messages:
+        audio_type: "can only be audio file"
+        audio_size: "maximum size is 5 Megabytes"
```
Comment (samnang) [inline config/locales/en.yml]: Change it to 10M

**Comment by samnang** [comment]
* [ ] Add province and communes in `callouts/show`
* [ ] Add province in `callouts/index`

**Pull request merged**